waiter/src/waiter/state.clj

```diff
@@ -79,44 +123,60 @@
+             (filter (fn [[_ {:keys [status-tags]}]]
+                       (if (and expired-instances? starting-instances?)
+                         (and (contains? status-tags :unhealthy)
+                              (not (contains? status-tags :starting)))
+                         true)))
+             (find-max instance-id-comparator)
+             first ; extract the instance-id
+             id->instance)))
 
 (defn find-available-instance
-  "For servicing requests, choose the oldest healthy instance with available slots."
+  "For servicing requests, choose the _oldest_ live healthy instance with available slots.
+   If such an instance does not exist, choose the _youngest_ expired healthy instance with available slots."
   [sorted-instance-ids id->instance instance-id->state acceptable-instance-id?]
-  (->> sorted-instance-ids
-       (filter acceptable-instance-id?)
-       (filter (fn [instance-id] (let [{:keys [status-tags] :as state} (get instance-id->state instance-id)]
-                                   (and (not (nil? state))
-                                        (slots-available? state)
-                                        (healthy? status-tags)))))
-       (first) ; extract the instance-id
-       (get id->instance)))
-
-(defn- instance-comparator
-  "The comparison order is: expired, service-id, started-at, and finally id."
-  [expired-instances]
-  (fn [x y]
-    (let [expired-comp (compare (contains? expired-instances (:id x)) (contains? expired-instances (:id y)))]
-      (if (zero? expired-comp)
-        (scheduler/instance-comparator x y)
-        expired-comp))))
+  (let [expired-instance? (fn [instance-id] (-> (instance-id->state instance-id) expired?))
+        order-instance-ids-based-on-expired-state (fn [instance-ids]
+                                                    (or (->> instance-ids
+                                                             (remove expired-instance?)
+                                                             seq)
+                                                        (->> instance-ids
+                                                             (filter expired-instance?)
+                                                             reverse
+                                                             seq)))]
+    (->> sorted-instance-ids
+         (filter acceptable-instance-id?)
+         (filter (fn [instance-id]
+                   (let [state (instance-id->state instance-id)]
+                     (and (not (nil? state))
+                          (slots-available? state)
+                          (healthy? state)))))
+         order-instance-ids-based-on-expired-state
+         first
+         id->instance)))
 
 (defn- sort-instances
-  "Sorts two service instances, the comparison order specified by `instance-comparator`."
+  "Sorts two service instances, the comparison order is: expired, service-id, started-at, and finally id."
```

Comment:
Changing this function also changes our instance distribution among routers?  Should we update how distribution works so that it doesn't depend on expiry?